main/ble_utils.c

```diff
@@ -298,7 +298,11 @@ char *chartoa(ble_uuid_t uuid, const uint8_t *data, size_t len)
             p += sprintf(p, "%s,", data[i] & 0x01 ? "true" : "false");
             break;
         case CHAR_TYPE_2BIT:
-            p += sprintf(p, "%" PRIu8 ",", data[i] & 0x03);
+            for(; i < len; i++){
+                for(int8_t j = 6; j >= 0; j -= 2){
+                    p += sprintf(p, "%" PRIu8 ",", (data[i] >> j) & 0b11);
```

Comment:
Why are you printing from the MSB first? Isn't the LSB the first digital IO?

Reply:
reversing the order could yield something like "0, 1, 1, 0, 3, 3, 0, 1" for 6 defined IOs, which I wanted to avoid.

Reply:
Is this from your implementation of this characteristic or a peripheral that passed the Bluetooth SIG certification?
I'd expect the LSB to be used first, especially since everything else in BLE is little endian

Reply:
Mine. And maybe I am wrong. I am just not aware of any other implementation out there.

Reply:
I'm not sure if I'm understanding it correctly either, but here's the spec: https://www.bluetooth.org/DocMan/handlers/DownloadDoc.ashx?doc_id=304972#page10

> The Digital characteristic contains the values of these bits in little endian order

Reply:
Little endian does not mean LSB or MSB

Reply:
True, but it doesn't make sense, to me, to read the bytes in little endian but then, to read the 2-bits from MSB to LSB but, read the 2-bit values in little endian again.
Also, since it's defined as a 2-bit array, I'd expect the array elements to follow each other in memory instead of jumping back and forth.

Reply:
I followed the natural reading direction, which  is left to right for me. If you insist we can change the direction though.

Reply:
Okay, it's LSB now